sdk/core/azure-core/azure/core/messaging.py

```diff
@@ -0,0 +1,168 @@
+        self.time = kwargs.pop("time", datetime.now(TZ_UTC))  # type: Optional[datetime]
+
+        self.datacontenttype = kwargs.pop("datacontenttype", None)  # type: Optional[str]
+        self.dataschema = kwargs.pop("dataschema", None)  # type: Optional[str]
+        self.subject = kwargs.pop("subject", None)  # type: Optional[str]
+        self.data = kwargs.pop("data", None)  # type: Optional[object]
+
+        try:
+            self.extensions = kwargs.pop("extensions")  # type: Optional[Dict]
+            for key in self.extensions.keys():  # type:ignore # extensions won't be None here
+                if not key.islower() or not key.isalnum():
+                    raise ValueError(
+                        "Extension attributes should be lower cased and alphanumeric."
+                    )
+        except KeyError:
+            self.extensions = None
+
+        if kwargs:
+            remaining = ", ".join(kwargs.keys())
+            raise ValueError(
+                "Unexpected keyword arguments {}. Any extension attributes must be passed explicitly using extensions."
+                .format(remaining)
+            )
+
+    def __repr__(self):
+        return "CloudEvent(source={}, type={}, specversion={}, id={}, time={})".format(
+            self.source, self.type, self.specversion, self.id, self.time
+        )[:1024]
+
+    @classmethod
+    def from_dict(cls, event):
+        # type: (Dict) -> CloudEvent
+        """
+        Returns the deserialized CloudEvent object when a dict is provided.
+        :param event: The dict representation of the event which needs to be deserialized.
+        :type event: dict
+        :rtype: CloudEvent
+        """
+        kwargs = {}  # type: Dict[Any, Any]
+        reserved_attr = [
+            "data",
+            "data_base64",
+            "id",
+            "source",
+            "type",
+            "specversion",
+            "time",
+            "dataschema",
+            "datacontenttype",
+            "subject",
+        ]
+
+        if "data" in event and "data_base64" in event:
+            raise ValueError(
+                "Invalid input. Only one of data and data_base64 must be present."
+            )
+
+        if "data" in event:
+            data = event.get("data")
+            kwargs["data"] = data if data is not None else NULL
+        elif "data_base64" in event:
+            kwargs["data"] = b64decode(
+                cast(Union[str, bytes], event.get("data_base64"))
+            )
+
+        for item in ["datacontenttype", "dataschema", "subject"]:
+            if item in event:
+                val = event.get(item)
+                kwargs[item] = val if val is not None else NULL
+
+        extensions = {k: v for k, v in event.items() if k not in reserved_attr}
+        if extensions:
+            kwargs["extensions"] = extensions
+
+        return cls(
+            id=event.get("id"),
+            source=cast(str, event.get("source")),
```

Comment:
If there is no `"source"` key in `event`, you will be (incorrectly) casting `None` to `str`. It is better to raise a `ValueError` here if the either of required keys (`"type"` and `"source"`) is not in `event` or else the error will surface much later.

Reply:
left it to event["source"]